### PR TITLE
[NOD-1207] Send reject messages

### PR DESCRIPTION
**Pull request by someone235**

### CODE REVIEW

File: netadapter/server/grpcserver/grpc_connection.go [hunk 1]
```diff
@@ -83,8 +84,13 @@ func (c *gRPCConnection) Disconnect() {
...
+		spawn("gRPCConnection.Disconnect-clientStream.CloseSend", func() {
+			// Wait a second before closing the stream, to let the send queue to get emptied.
+			const finishSendDuration = time.Second
+			time.Sleep(finishSendDuration)
+			clientStream := c.stream.(protowire.P2P_MessageStreamClient)
+			_ = clientStream.CloseSend() // ignore error because we don't really know what's the status of the connection
+		})
 	}
 
 	log.Debugf("Disconnected from %s", c)
```
Comment (stasatdaglabs) [inline netadapter/server/grpcserver/grpc_connection.go:96]: Please move this log into the goroutine above and consider adding a "Disconnecting..." log.

File: netadapter/server/grpcserver/protowire/messages.proto [hunk 2]
```diff
@@ -212,6 +213,12 @@ message VersionMessage{
 }
 // VersionMessage end
 
+// RejectMessage start
+message RejectMessage{
+  string reason = 2;
```
Comment (stasatdaglabs) [inline netadapter/server/grpcserver/protowire/messages.proto:218]: Please change this to = 1.

File: protocol/flows/rejects/handle_rejects.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+type RejectsContext interface {
+}
+
+type handleRejectsFlow struct {
+	RejectsContext
+	incomingRoute, outgoingRoute *router.Route
+}
+
+// ReceivePings handles all ping messages coming through incomingRoute.
+// This function assumes that incomingRoute will only return MsgPing.
```
Comment (stasatdaglabs) [inline protocol/flows/rejects/handle_rejects.go:19]: This comment is wrong. Please correct it.

File: protocol/protocol.go [hunk 4]
```diff
@@ -58,28 +59,29 @@ func (m *Manager) routerInitializer(router *routerpkg.Router, netConnection *net
...
-func (m *Manager) handleError(err error, netConnection *netadapter.NetConnection) {
+func (m *Manager) handleError(err error, netConnection *netadapter.NetConnection, outgoingRoute *routerpkg.Route) {
 	if protocolErr := &(protocolerrors.ProtocolError{}); errors.As(err, &protocolErr) {
 		if !m.context.Config().DisableBanning && protocolErr.ShouldBan {
 			log.Warnf("Banning %s (reason: %s)", netConnection, protocolErr.Cause)
 			err := m.context.ConnectionManager().Ban(netConnection)
 			if err != nil && !errors.Is(err, addressmanager.ErrAddressNotFound) {
 				panic(err)
 			}
+			_ = outgoingRoute.Enqueue(domainmessage.NewMsgReject(protocolErr.Error()))
```
Comment (stasatdaglabs) [inline protocol/protocol.go:84]: Please check RouteIsClosed explicitly.

File: protocol/protocol.go [hunk 5]
```diff
@@ -216,6 +218,19 @@ func (m *Manager) registerTransactionRelayFlow(router *routerpkg.Router, isStopp
 	}
 }
 
+func (m *Manager) registerRejectsFlow(router *routerpkg.Router, isStopping *uint32, errChan chan error) []*flow {
```
Comment (stasatdaglabs) [inline protocol/protocol.go:221]: It appears this isn't called from anywhere.

File: netadapter/router/route.go [hunk 6]
```diff
@@ -101,3 +101,8 @@ func (r *Route) Close() {
 	r.closed = true
 	close(r.channel)
 }
+
+// IsEmpty returns whether the route has any pending messages
```
Comment (svarogg) [inline netadapter/router/route.go:105]: I prefer the format "IsEmpty returns true/false if..."

Usually the format "XXX returns whether YYY" implies XXX returns true if YYY.
In this case it's the other way around.

"XXX returns YYY if ZZZ" is doesn't have this ambiguity.

File: netadapter/server/grpcserver/connection_loops.go [hunk 7]
```diff
@@ -33,7 +33,8 @@ func (c *gRPCConnection) connectionLoops() error {
 
 func (c *gRPCConnection) sendLoop() error {
 	outgoingRoute := c.router.OutgoingRoute()
-	for c.IsConnected() {
+	// Once the connected is closed, the send loop empties the remaining messages in the channel.
```
Comment (svarogg) [inline netadapter/server/grpcserver/connection_loops.go:36]: Typo: Once the **connection** is closed...

File: domainmessage/msgreject.go [hunk 8]
```diff
@@ -0,0 +1,22 @@
+package domainmessage
+
+// MsgReject implements the Message interface and represents a kaspa
+// Reject message. It is used to notify peers why they are banned.
+type MsgReject struct {
+	baseMessage
+	Reason string
```
Comment (svarogg) [inline domainmessage/msgreject.go:7]: Please make sure the reject message contains the type (Block/Transaction) and Hash/ID of what failed.

I'd rather have it as two separate fields in the message, but if it's already in the Reason string, I don't mind keeping it this way for now.

**Pull request merged**